### ISSUE TITLE
unable to recreate results of self-feeding chatbot paper

### ISSUE THREAD

**Issue by ekunnii** [facebookresearch/ParlAI#2452]
**Bug description**
For dialog agent with human to human dataset, I ran the training for 9 rounds and got the higher number than results on paper, 46.25 vs 44.7. 

For diafee agent with human to human and feedback data, I ran the training for 14 rounds and got the lower number than results on paper.

This invalidates the claim of paper that with feedback data, chatbot should performs better.

  |   | Valid(dia_acc) | Test(dia_acc)
-- | -- | -- | --
1 | Chatbot hh paper | NA | 44.7 (0.4)
2 | Chatbot hh+fb paper | NA | 45.8 (0.7)
3 | Chatbot with hh reproduce | 44.96(1.38) | 46.25(1.42)
4 | Chatbot with hh+fb reproduce | 43.82(1.36) | 45.07(1.3)

**Reproduction steps**
These are training commands:
```
python examples/train_model.py -t self_feeding:dialog --model-file /data/dialog_baseline_`date +%m%d_%H%M%S` -ltim 5 -vtim 10 -vp 50 -m projects.self_feeding.self_feeding_agent:SelfFeedingAgent -cands batch --eval-candidates inline -histsz 2 --embedding-type fasttext_cc --embedding-size 300 --dict-maxtokens 250000 --num-epochs 500 --optimizer adamax --embeddings-scale false -bs 128 --relu-dropout 0 --attention-dropout 0 --n-heads 2 --n-layers 2 -lr 0.0025 --ffn-size 32 --lr-scheduler invsqrt --warmup-updates 500 -vmt dia_acc -vmm max &> /data/dialog_baseline_`date +%m%d_%H%M%S`.log &

python examples/train_model.py -t self_feeding:diafee --model-file /data/diafee_baseline_`date +%m%d_%H%M%S` -ltim 5 -vtim 10 -vp 50 -m projects.self_feeding.self_feeding_agent:SelfFeedingAgent -cands batch --eval-candidates inline -histsz 2 --embedding-type fasttext_cc --embedding-size 300 --dict-maxtokens 250000 --num-epochs 500 --optimizer adamax --embeddings-scale false -bs 128 --relu-dropout 0 --attention-dropout 0 --n-heads 2 --n-layers 2 -lr 0.0025 --ffn-size 32 --lr-scheduler invsqrt --warmup-updates 500 -vmt dia_acc -vmm max &> /data/diafee_baseline_`date +%m%d_%H%M%S`.log &
```
**Expected behavior**
By adding feedback data, accuracy should be higher than model without any feedback data

**Comment by stephenroller** [comment]
We've made a lot of changes that might be breaking something. Maybe you can try one of these older commits:

https://github.com/facebookresearch/ParlAI/commits/master/projects/self_feeding

I'd recommend 971b12b2838d55cd0a98ede502946ef9ccbe04c6.

**Comment by ekunnii** [comment]
Thanks Stephen.

I ran four training on suggested commit id. However, the results are much lower than expected.
test accuracy on dialog agent is 24.01(1.99)

So I decided to roll back further to commit id  0602a00  and ran four training tasks. The results are higher than expected this time, which is 50.3775(0.29). I choose this commit as it has bug fixes and before the refactor of utils. 

I also tried the initial commit  e43354a . But training script failed with errors cannot find the module of self-feeding. 

Can you suggest how to move forward? Thanks.

I always use same set of hyper-parameters. Here is the command I used to start training.
```
python examples/train_model.py -t self_feeding:dialog --model-file /data/dialog_baseline_`date +%m%d_%H%M%S` -ltim 5 -vtim 10 -vp 50 -m projects.self_feeding.self_feeding_agent:SelfFeedingAgent -cands batch --eval-candidates inline -histsz 2 --embedding-type fasttext_cc --embedding-size 300 --dict-maxtokens 250000 --num-epochs 500 --optimizer adamax --embeddings-scale false -bs 128 --relu-dropout 0 --attention-dropout 0 --n-heads 2 --n-layers 2 -lr 0.0025 --ffn-size 32 --lr-scheduler invsqrt --warmup-updates 500 -vmt dia_acc -vmm max &> /data/dialog_baseline_`date +%m%d_%H%M%S`.log &
```

**Comment by stephenroller** [comment]
Great, thanks for all the thorough attempts. I will take a look

**Comment by ekunnii** [comment]
Thanks Stephen. 
In my latest attempt, I check out the commit db90986, which is the last commit from Braden Hancock. 

With that commit, I am still not able to recreate the results. However, this time, the result validates the claim that feedback data helps to  improve chatbot performance. But the improvement is not as large as ones reported on paper.

  |   | Valid(dia_acc) | Test(dia_acc)
-- | -- | -- | --
1 | Chatbot hh paper | NA | 44.7 (0.4)
2 | Chatbot hh fb paper | NA | 45.8 (0.7)
3 | Chatbot with hh reproduce | 44.96(1.38) | 46.25(1.42)
4 | Chatbot with hh+fb reproduce | 43.82(1.36) | 45.07(1.3)
7 | Chatbot with hh reproduce(commit db90986) | 49.025(0.66) | 49.49(0.49)
8 | Chatbot with hh+fb reproduce (commit db90986) | 49.27(1.06) | 49.97(1.3)

**Comment by stephenroller** [comment]
Thanks for the more details. What are the parens? Are they variances?

@bhancock8 you don't have any knowledge do you?

**Comment by bhancock8** [comment]
Hi @ekunnii, thanks for the detailed tables! I can't say for sure without more context, but one thing I don't see in your training commands is pointers to the specific train files you're using (see the [README](https://github.com/facebookresearch/ParlAI/tree/master/projects/self_feeding) for examples of how to specify these). Can you confirm which data files you're training on?

**Comment by ekunnii** [comment]
Hello @bhancock8, Thanks for reply. I am reproducing dialog agent with HH only and diafee with HH and FB. According to sample commands on README, I don't have to specify the train file for dialog or diafee. Besides that, as shown in training log, I can confirm that correct files had been loaded.
```
[creating task(s): self_feeding:dialog]
[ Loading Self-Feeding text data:/home/ryan/ParlAI/data/self_feeding/train_hh.txt]
```
```
[creating task(s): self_feeding:diafee]
[ Loading Self-Feeding text data:/home/ryan/ParlAI/data/self_feeding/train_hh.txt]
[ Loading Self-Feeding text data:/home/ryan/ParlAI/data/self_feeding/train_fb.txt]
```
I have another question regarding vp and epochs. When training dialog agent, '-vp 10' and '--num-epochs 100' lead to very bad result. Test accuracy is only 14.4875(1.87) , So I align them with full model parameter values which are  '-vp 50' and '--num-epochs 500' and then got the above results.
Do you remember the reason why different values are used for training dialog, diafee and full model? Thanks.

**Comment by stephenroller** [comment]
That's validation patience (for early stopping) and max training epochs. If -vp 10 is giving bad results, then it is learning extremely slowly.

**Comment by bhancock8** [comment]
@ekunnii, if you're just using the default files, then you're training on 131k HH examples and 131k HB examples; the default HB file, `train_hb.txt`, contains all available examples. So if you're comparing to the numbers in Table 3, you'll want to pull out just 60k of those to create train_hb60k.txt and point to that file when training (we describe the dataset sizes and give examples commands for what I just described in the [project README](https://github.com/facebookresearch/ParlAI/tree/master/projects/self_feeding)). It very well may be the case that for this particular dataset, when the number of HB examples is the same size as the number of HH examples, the expected lift reduces. Perhaps in that situation, one would be better off pre-training on the HB examples before fine-tuning on HH. In general, I find that question of when to pre-train/fine-tune vs. do multitask learning a very interesting one, and one that I hope the field will continue to explore and try to understand!

One other thing to consider:
- The numbers reported in the paper (Table 3) are averaged over 20 runs, and include standard deviations up to 0.8. At n=20, the gaps are significant (see the caption of that Table for details), but on individual runs we can expect the absolute differential to vary.
